### ISSUE TITLE
Accessing any WebApp URL causes infinite 404 redirect loop

### ISSUE THREAD

**Issue by kblaschke** [mattermost/mattermost-server#10671]
#### Summary
Mattermost WebApp infinitely redirects to /error URL because all URLs raise a 404 Not Found error, including the error page.

#### Steps to reproduce
Using version 5.10.0. It is quite unclear how to reproduce the behaviour, but it seems to be related to a database/update issue. The only error logged on startup is:

```ERROR:  relation "idx_teams_description" does not exist```

The same error has been [posted in the community forums](https://forum.mattermost.org/t/redirect-loop-sorry-we-could-not-find-the-page/5033) about a year ago, so the issue is a) not new and b) still not fixed in the current version.
The fix working for the user who reported the issue in the forums was to delete the whole database, but this is clearly not an option.

The Mattermost server is running an an Alpine Linux Docker container with a - mostly - default configuration. The reverse proxy is not an issue here, even a ```curl http://localhost:8000``` inside the container returns the same error page.

The subpath URL is "/", the server is running without any prefix (only on a subdomain, which is correctly configured in ServiceSettings.SiteURL).

#### Expected behavior
When accessing the Mattermost Webapp base URL with a browser or through the Mattermost desktop app, the login page should be shown.

#### Observed behavior (that appears unintentional) 
Accessing any URL that doesn't start with /api/ or /hooks/ (and thus is handled as a Webapp page) redirects to ```/error?message=Sorry%2C+we+could+not+find+the+page.&s=(session ID)```. As the /error URL also cannot be found, this leads to a redirect loop, until the rate limiter of the Mattermost server kicks in and blocks the client.

**Comment by kblaschke** [comment]
Some further information that may help, I debugged the request with Delve.
Stack:
```
0  0x0000000000eee982 in github.com/mattermost/mattermost-server/app.(*App).Handle404
   at /go/src/github.com/mattermost/mattermost-server/app/app.go:135
1  0x00000000010176a8 in github.com/mattermost/mattermost-server/app.(*App).Handle404-fm
   at /go/src/github.com/mattermost/mattermost-server/app/app.go:134
2  0x00000000006e63e4 in net/http.HandlerFunc.ServeHTTP
   at /usr/local/go/src/net/http/server.go:1995
3  0x0000000000d3227c in github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.(*Router).ServeHTTP
   at /go/src/github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux/mux.go:162
4  0x000000000100d638 in github.com/mattermost/mattermost-server/app.(*RateLimiter).RateLimitHandler.func1
   at /go/src/github.com/mattermost/mattermost-server/app/ratelimit.go:107
5  0x00000000006e63e4 in net/http.HandlerFunc.ServeHTTP
   at /usr/local/go/src/net/http/server.go:1995
6  0x00000000006e979b in net/http.serverHandler.ServeHTTP
   at /usr/local/go/src/net/http/server.go:2774
7  0x00000000006e53ac in net/http.(*conn).serve
   at /usr/local/go/src/net/http/server.go:1878
8  0x000000000045e551 in runtime.goexit
   at /usr/local/go/src/runtime/asm_amd64.s:1337
```
Locals in mux.ServeHTTP:
```
(dlv) p req
*net/http.Request {
        Method: "GET",
        URL: *net/url.URL {
                Scheme: "",
                Opaque: "",
                User: *net/url.Userinfo nil,
                Host: "",
                Path: "/",
                RawPath: "",
                ForceQuery: false,
                RawQuery: "",
                Fragment: "",},
        Proto: "HTTP/1.1",
        ProtoMajor: 1,
        ProtoMinor: 1,
        Header: net/http.Header [
                "": nil, 
                "User-Agent": [
                        "curl/7.64.0",
                ], 
        ],
        Body: io.ReadCloser(net/http.noBody) {},
        GetBody: nil,
        ContentLength: 0,
        TransferEncoding: []string len: 0, cap: 0, nil,
        Close: false,
        Host: "localhost:8000",
        Form: net/url.Values nil,
        PostForm: net/url.Values nil,
        MultipartForm: *mime/multipart.Form nil,
        Trailer: net/http.Header nil,
        RemoteAddr: "127.0.0.1:55680",
        RequestURI: "/",
        TLS: *crypto/tls.ConnectionState nil,
        Cancel: <-chan struct {} {},
        Response: *net/http.Response nil,
        ctx: context.Context(*context.valueCtx) *{
                Context: context.Context(*context.valueCtx) ...,
                key: interface {}(github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.contextKey) *(*interface {})(0xc001984790),
                val: interface {}(*github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route) ...,},}
(dlv) p match
github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.RouteMatch {
        Route: *github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route {
                parent: github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.parentRoute(*github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Router) ...,
                handler: net/http.Handler nil,
                matchers: []github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.matcher len: 2, cap: 2, [
                        ...,
                        ...,
                ],
                regexp: *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.routeRegexpGroup")(0xc00116aab0),
                strictSlash: false,
                skipClean: false,
                useEncodedPath: false,
                buildScheme: "",
                buildOnly: false,
                name: "",
                err: error nil,
                buildVarsFunc: nil,},
        Handler: net/http.Handler(net/http.HandlerFunc) github.com/mattermost/mattermost-server/app.(*App).Handle404-fm,
        Vars: map[string]string [],
        MatchErr: error(*errors.errorString) *{
                s: "no matching route was found",},}
(dlv) p match.Route.matchers
[]github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.matcher len: 2, cap: 2, [
        *github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.routeRegexp {
                template: "/",
                regexpType: regexpTypePrefix (2),
                options: (*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.routeRegexpOptions")(0xc000122718),
                regexp: *(*regexp.Regexp)(0xc001555950),
                reverse: "/",
                varsN: []string len: 0, cap: 0, [],
                varsR: []*regexp.Regexp len: 0, cap: 0, [],},
        *github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Router {
                NotFoundHandler: net/http.Handler(net/http.HandlerFunc) *(*"net/http.Handler")(0xc000122770),
                MethodNotAllowedHandler: net/http.Handler nil,
                parent: github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.parentRoute(*github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route) ...,
                routes: []*github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route len: 20, cap: 32, [
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc0015559e0),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc001555c20),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc001554900),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc003f3b8c0),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc002436a20),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc002436b40),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc002436c60),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc002436d80),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc002436ea0),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc0024370e0),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc002437320),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc002437560),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc0024377a0),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc0024379e0),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc002437c20),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc00249ccf0),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc00249cf30),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc00249d170),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc00249d3b0),
                        *(*"github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route")(0xc00249d4d0),
                ],
                namedRoutes: map[string]*github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.Route nil,
                strictSlash: false,
                skipClean: false,
                KeepContext: false,
                useEncodedPath: false,
                middlewares: []github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.middleware len: 0, cap: 0, nil,},
]
```

**Comment by kblaschke** [comment]
After some additional debugging, I found out that the setting "ServiceSettings.WebserverMode" was set to "disabled". In that case, web.InitStatic() doesn't register any routes, not even for / and /error, which then triggers the redirect loop.

IMHO, two possible solutions to this annoying problem are possible. Either remove the 404 error redirect and just return a blank page if WebserverMode is set to "disabled", or leave at least the /error handler so a proper error page can be displayed.

Also, the [documentation of this setting](https://docs.mattermost.com/administration/config-settings.html#webserver-mode) should be fixed. It just says it controls whether the service will serve static files, but this is misleading. If set to "disabled", it will stop serving the whole Webapp (including the Desktop apps), not just some static content.

**Comment by amyblais** [comment]
@svelle Do you know about this if there is a bug here?

**Comment by amyblais** [comment]
Opened a ticket here: https://mattermost.atlassian.net/browse/MM-15367.

**Comment by amyblais** [comment]
Closing this as this has been fixed for the upcoming v5.12 release (to be released on June 16th).
